crypto/evp/evp_fetch.c

```diff
@@ -0,0 +1,182 @@
+};
+
+/*
+ * Generic routines to fetch / create EVP methods with ossl_method_construct()
+ */
+static void *alloc_tmp_method_store(void)
+{
+    return ossl_method_store_new();
+}
+
+ static void dealloc_tmp_method_store(void *store)
+{
+    if (store != NULL)
+        ossl_method_store_free(store);
+}
+
+static
+struct OSSL_METHOD_STORE *get_default_method_store(OPENSSL_CTX *libctx)
+{
+    if (!RUN_ONCE(&default_method_store_init_flag,
+                  do_default_method_store_init))
+        return NULL;
+    return openssl_ctx_get_data(libctx, default_method_store_index);
+}
+
+static void *get_method_from_store(OPENSSL_CTX *libctx, void *store,
+                                   const char *propquery, void *data)
+{
+    struct method_data_st *methdata = data;
+    void *method = NULL;
+
+    if (store == NULL
+        && (store = get_default_method_store(libctx)) == NULL)
+        return NULL;
+
+    (void)ossl_method_store_fetch(store, methdata->nid, propquery, &method);
+
+    return method;
+}
+
+static int put_method_in_store(OPENSSL_CTX *libctx, void *store,
+                               const char *propdef, void *method,
+                               void *data)
+{
+    struct method_data_st *methdata = data;
+    int ret = 0;
+
+    if (store == NULL
+        && (store = get_default_method_store(libctx)) == NULL)
+        return 0;
+
+    return ossl_method_store_add(store, methdata->nid, propdef, method,
+                                 methdata->mcm->destruct);
+}
+
+static void *construct_method(const OSSL_DISPATCH *fns, OSSL_PROVIDER *prov,
+                              void *data)
+{
+    struct method_data_st *methdata = data;
+    void *method = NULL;
+
+    if (methdata->nid == NID_undef) {
+        /* Create a new NID for that name on the fly */
+        ASN1_OBJECT tmpobj;
+
+        /* This is the same as OBJ_create() but without requiring a OID */
+        tmpobj.nid = OBJ_new_nid(1);
+        tmpobj.sn = tmpobj.ln = methdata->name;
+        tmpobj.flags = ASN1_OBJECT_FLAG_DYNAMIC;
+        tmpobj.length = 0;
+        tmpobj.data = NULL;
+
+        methdata->nid = OBJ_add_object(&tmpobj);
+    }
+
+    if (methdata->nid == NID_undef)
+        return NULL;
+
+    method = methdata->method_from_dispatch(methdata->nid, fns, prov);
+    if (method == NULL)
+        return NULL;
+    return method;
+}
+
+void *evp_generic_fetch(OPENSSL_CTX *libctx, int operation_id,
+                        const char *algorithm, const char *properties,
+                        void *(*new_method)(int nid, const OSSL_DISPATCH *fns,
+                                            OSSL_PROVIDER *prov),
+                        int (*upref_method)(void *),
+                        void (*free_method)(void *))
+{
+    int nid = OBJ_sn2nid(algorithm);
+    void *method = NULL;
+
+    if (nid != NID_undef
+        && ossl_method_store_cache_get(NULL, nid, properties,
+                                       (void **)&method)) {
```

Comment:
Because compilers are "helpfully" warning about the type loss otherwise...  it's the same kind of helpfulness as when you try to pass a `char **` to a `const char **` parameter.  It's really annoying, but there you go...

Reply:
I must be being stupid...what type loss are you referring to? I don't see any.

Reply:
Looks like you're right...  this must be an artefact from older versions of the code that didn't get cleaned away.  So, er, thanks :-)

Reply:
Fixed